kubespawner/spawner.py

```diff
@@ -173,8 +173,12 @@ def __init__(self, *args, **kwargs):
         # By now, all the traitlets have been set, so we can use them to
         # compute other attributes
 
+        self.ns_labels = {}
+        self.ns_annotations = {}
         if self.enable_user_namespaces:
             self.namespace = self._expand_user_properties(self.user_namespace_template)
+            self.ns_labels = self.user_namespace_labels
+            self.ns_annotations = self.user_namespace_annotations
```

Comment:
To allow expansion of user properties could be relevant, but hmmmm you can leave it out but a comment should be made in the docstring about the state of how this is implemented - will an expansion be made or not?

I guess, the use of `_template` indicated that expansion will take place, and that it really must include something to expand to be username unique.

If you are not expanding things, I suggest removing the intermediary variables `ns_labels` and instead relying on the configuration directly.

Reply:
Hmm, now that I think about it expansion could be very useful, I haven't thought of including that. I did the `ns_labels` and `ns_annotations` mainly because I thought it would look better than having an `if` statement in `_ensure_namespace`.

I think that the expansion could work by iterating through the configuration dictionary and for each key assign expanded value to `ns_labels` or `ns_annotations` under the same key.

Do you think this solution would be acceptable?

Reply:
I've added expansion through the `_expand_all` function and use it when creating the namespace, so this should take care of the possibility of different logic based on users as you mentioned.